pkg/githubreconciler/issuemanager/session.go

```diff
@@ -0,0 +1,303 @@
+/*
+Copyright 2025 Chainguard, Inc.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package issuemanager
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/chainguard-dev/clog"
+	"github.com/chainguard-dev/terraform-infra-common/pkg/githubreconciler"
+	"github.com/google/go-github/v75/github"
+)
+
+// IssueSession represents work on multiple issues for a specific resource path.
+// Unlike Session in changemanager which handles a single PR, IssueSession manages multiple issues.
+// T must implement the Comparable interface to enable matching between existing and desired issues.
+type IssueSession[T Comparable[T]] struct {
+	manager        *IM[T]
+	client         *github.Client
+	resource       *githubreconciler.Resource
+	pathLabel      string
+	existingIssues []*github.Issue
+}
+
+// hasSkipLabel checks if a specific issue has the skip label.
+// Returns true if the issue has a label matching "skip:{identity}".
+func (s *IssueSession[T]) hasSkipLabel(issue *github.Issue) bool {
+	skipLabel := "skip:" + s.manager.identity
+	for _, label := range issue.Labels {
+		if label.GetName() == skipLabel {
+			return true
+		}
+	}
+	return false
+}
+
+// UpsertMany creates or updates multiple issues based on the provided data.
+// It uses the Equal method on T to determine which existing issues correspond to which desired data.
+// The pathLabel is automatically added to the provided labels.
+// Returns a slice of issue URLs in the same order as the input data.
+func (s *IssueSession[T]) UpsertMany(
+	ctx context.Context,
+	desired []*T,
+	labels []string,
+) ([]string, error) {
+	log := clog.FromContext(ctx)
+
+	// Add pathLabel to labels
+	allLabels := append([]string{s.pathLabel}, labels...)
+
+	issueURLs := make([]string, len(desired))
+
+	for i, data := range desired {
+		// Try to find matching existing issue
+		existingIssue := s.findMatchingIssue(ctx, data)
+
+		if existingIssue != nil {
+			// Check if issue has skip label
+			if s.hasSkipLabel(existingIssue) {
+				log.Infof("Issue #%d has skip label, skipping update", existingIssue.GetNumber())
+				issueURLs[i] = existingIssue.GetHTMLURL()
+				continue
+			}
+
+			// Check if update is needed
+			needsUpdate, err := s.needsUpdate(ctx, existingIssue, data)
```

Comment:
Just to illustrate my point re: repeated parsing.

In the loop over desired we parse potentially ALL of the existing issues each time in `findMatchingIssue`, then again to check `needsUpdate`, and then (probably) again in `CloseAnyOutstanding`.